### PR TITLE
koord-scheduler: Add non-preemptible feature in Elastic Quota Management

### LINKED ISSUE #1473: [proposal] Add non-preemptible feature in Elastic Quota Management
**What is your proposal**:
- Background
  - In the Google Borg paper, it describes a feature that "production" workloads cannot be preempted.
  > Although a preempted task will often be rescheduled elsewhere in the cell, preemption cascades could occur if a high-priority task bumped out a slightly lower-priority one, which bumped out another slightly-lower priority task, and so on. To eliminate most of this, **we disallow tasks in the production priority band to preempt one another.**
  - As [Multi Hierarchy Elastic Quota Management](https://koordinator.sh/docs/designs/multi-hierarchy-elastic-quota-management) documented, if the workload under a ElasticQuota uses more than "Min" resources, regardless of its Priority/QoS level, the workload may be preempted.

- Proposal
  - For important workloads, can we support setting Pods as non-preemptible (e.g. set koordinator.sh/preemptible: false), **so the sum of resource usage of Pods with this setting cannot exceed the Min of ElasticQuota, thus avoiding preemption** (Or we can limit "koord-prod" Priority workloads' resource usage not exceeding Min by configuration).

- Reference
  - [Uber Peloton](https://peloton.readthedocs.io/en/latest/user-guide/#resource-pools).

**Why is this needed**:
With this feature, users can deploy different types of workloads in a single ElasticQuota, including production workloads which cannot tolerate preemption, and non-production workloads which can take the advantages of ElasticQuota for more resources.

**Is there a suggested solution, if so, please add it**:

- Solution A: Add a new label (e.g. koordinator.sh/preemptible: true/false) to control whether a Pod can be preempted or not.
- Solution B: Add a configuration to control whether Pods with "koord-prod" Priority can be preempted or not.

Both Solution A and B should ensure that the sum of resource usage of non-preemptible Pods cannot exceed the Min of ElasticQuota.

### PR DESCRIPTION
**Pull request by tan90github**
### Ⅰ. Describe what this PR does

Add non-preemptible feature in Elastic Quota Management
### Ⅱ. Does this pull request fix one issue?

fixes #1473
### V. Checklist

- [ ] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/scheduler/plugins/elasticquota/plugin_test.go [hunk 1]
```diff
@@ -750,6 +750,123 @@ func TestPlugin_PreFilter_CheckParent(t *testing.T) {
...
+			for _, qis := range tt.quotaInfos {
+				gp.OnQuotaAdd(qis)
+			}
+			for _, pod := range tt.initPods {
+				gp.OnPodAdd(pod)
+			}
+
+			qi := gp.groupQuotaManager.GetQuotaInfoByName("test1")
+			qi.Lock()
+			qi.CalculateInfo.Runtime = tt.runtime.DeepCopy()
```
Comment (shaloulcy) [inline pkg/scheduler/plugins/elasticquota/plugin_test.go:859]: runtime is not need to seet

File: pkg/scheduler/plugins/elasticquota/plugin_test.go [hunk 2]
```diff
@@ -750,6 +750,123 @@ func TestPlugin_PreFilter_CheckParent(t *testing.T) {
...
+			qi.UnLock()
+			fmt.Println(len(qi.GetPodCache()))
+			state := framework.NewCycleState()
+			ctx := context.TODO()
+			_, status := gp.PreFilter(ctx, state, tt.pod)
```
Comment (shaloulcy) [inline pkg/scheduler/plugins/elasticquota/plugin_test.go:864]: tt.pod should has no nodeName.

File: pkg/scheduler/plugins/elasticquota/plugin.go [hunk 3]
```diff
@@ -226,12 +229,22 @@ func (g *Plugin) PreFilter(ctx context.Context, cycleState *framework.CycleState
 
 	podRequest, _ := core.PodRequestsAndLimits(pod)
 	used := quotav1.Add(podRequest, state.used)
+	if extension.IsPodNonPreemptible(pod) {
+		nonPreemptibleUsed = state.nonPreemptibleUsed
+		addNonPreemptibleUsed = quotav1.Add(podRequest, nonPreemptibleUsed)
+	}
```
Comment (eahydra) [inline pkg/scheduler/plugins/elasticquota/plugin.go:235]: move the codes to line 242

File: pkg/scheduler/plugins/elasticquota/plugin.go [hunk 4]
```diff
@@ -207,6 +209,7 @@ func (g *Plugin) EventsToRegister() []framework.ClusterEvent {
 }
 
 func (g *Plugin) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod) (*framework.PreFilterResult, *framework.Status) {
+	var nonPreemptibleUsed, addNonPreemptibleUsed = make(corev1.ResourceList), make(corev1.ResourceList)
```
Comment (eahydra) [inline pkg/scheduler/plugins/elasticquota/plugin.go:212]: Invalid code.. in line 233 and 234, got new values.

File: pkg/scheduler/plugins/elasticquota/plugin.go [hunk 5]
```diff
@@ -226,12 +229,22 @@ func (g *Plugin) PreFilter(ctx context.Context, cycleState *framework.CycleState
...
+	}
 
 	if isLessEqual, exceedDimensions := quotav1.LessThanOrEqual(used, state.runtime); !isLessEqual {
 		return nil, framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Insufficient quotas, "+
 			"quotaName: %v, runtime: %v, used: %v, pod's request: %v, exceedDimensions: %v",
 			quotaName, printResourceList(state.runtime), printResourceList(state.used), printResourceList(podRequest), exceedDimensions))
 	}
+	quotaMin := state.quotaInfo.CalculateInfo.Min
+	if isLessEqual, exceedDimensions := quotav1.LessThanOrEqual(addNonPreemptibleUsed, quotaMin); !isLessEqual {
+		return nil, framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Insufficient non-preemptible quotas, "+
```
Comment (eahydra) [inline pkg/scheduler/plugins/elasticquota/plugin.go:244]: The message is clear. 👍🏻

File: apis/extension/elastic_quota.go [hunk 6]
```diff
@@ -38,6 +38,7 @@ const (
 	LabelQuotaProfile         = QuotaKoordinatorPrefix + "/profile"
 	LabelQuotaIsRoot          = QuotaKoordinatorPrefix + "/is-root"
 	LabelQuotaTreeID          = QuotaKoordinatorPrefix + "/tree-id"
+	LabelNonPreemptible       = QuotaKoordinatorPrefix + "/non-preemptible"
```
[the remark this answers is not in the document]
Comment (hormes) [reply apis/extension/elastic_quota.go]: I think it's better to put it under quota. The semantics are clear. Regarding the priority preemption behavior, whether it is necessary to support similar semantics, please keep an eye out for the time being.

**Comment by hormes** [comment]
After the non-preemptible Pod reaches min, should it be rejected now or should it remain pending?

**Comment by shaloulcy** [comment]
> After the non-preemptible Pod reaches min, should it be rejected now or should it remain pending?

pending

**Comment by tan90github** [comment]
> After the non-preemptible Pod reaches min, should it be rejected now or should it remain pending?

now it will pending

File: apis/extension/elastic_quota.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (eahydra) [reply apis/extension/elastic_quota.go]: OK.. As I said before, maybe in the future when we need to implement a new preemption process, we may need to define a new preemptible api, and at the same time, we must also take into account the non-preemptible api in the Quota mechanism.

File: pkg/scheduler/plugins/elasticquota/core/group_quota_manager.go [hunk 7]
```diff
@@ -615,28 +616,35 @@ func (gqm *GroupQuotaManager) updatePodUsedNoLock(quotaName string, oldPod, newP
...
 	deltaUsed := quotav1.Subtract(newPodUsed, oldPodUsed)
-	if quotav1.IsZero(deltaUsed) {
-		if klog.V(5).Enabled() {
-			klog.Infof("updatePodUsed, deltaUsedIsZero, quotaName: %v, podName: %v, podUsed: %v",
-				quotaName, getPodName(oldPod, newPod), util.DumpJSON(newPodUsed))
-		}
+	deltaNonPreemptibleUsed := quotav1.Subtract(newNonPreemptibleUsed, oldNonPreemptibleUsed)
+	if quotav1.IsZero(deltaUsed) && quotav1.IsZero(deltaNonPreemptibleUsed) {
+		klog.V(5).Infof("updatePodUsed, deltaUsed and deltaNonPreemptibleUsed IsZero, quotaName:%v, podName:%v, podUsed:%v, podNonPreemptibleUsed:%v",
+			quotaName, getPodName(oldPod, newPod), newPodUsed, newNonPreemptibleUsed)
```
Comment (shaloulcy) [inline pkg/scheduler/plugins/elasticquota/core/group_quota_manager.go:644]: resume util.DumpJSON. The log will more readability

**Review by eahydra — APPROVED** [review]
/lgtm
/approve

**Review by eahydra — COMMENTED** [review]
/lgtm

**Pull request merged**